src/log/Entry.h

```diff
@@ -4,103 +4,113 @@
-      m_next(NULL),
-      m_buf_len(buf_len),
-      m_exp_len(exp_len),
-      m_data(buf, buf_len),
-      m_streambuf(CachedPrebufferedStreambuf::create(&m_data))
-  {
-    if (msg) {
-      get_ostream() << msg;
-    }
-  }
+  Entry(const Entry &) = default;
+  Entry& operator=(const Entry &) = default;
+  Entry(Entry &&e) = default;
+  Entry& operator=(Entry &&e) = default;
+  virtual ~Entry() = default;
+
+  virtual std::string_view strv() const = 0;
+  virtual std::size_t size() const = 0;
+
+  time m_stamp;
+  pthread_t m_thread;
+  short m_prio, m_subsys;
 
 private:
-  ~Entry() = default;
+  log_clock& clock() {
+    static log_clock clock;
+    return clock;
+  }
+};
 
+/* This should never be moved to the heap! Only allocate this on the stack. See
+ * CachedStackStringStream for rationale.
+ */
+class MutableEntry : public Entry {
 public:
+  MutableEntry() = delete;
+  MutableEntry(short pr, short sub) : Entry(pr, sub) {}
+  MutableEntry(const MutableEntry&) = delete;
+  MutableEntry& operator=(const MutableEntry&) = delete;
+  MutableEntry(MutableEntry&&) = delete;
+  MutableEntry& operator=(MutableEntry&&) = delete;
+  ~MutableEntry() override = default;
+
   std::ostream& get_ostream() {
-    return m_streambuf->get_ostream();
+    return cos.get_stream();
   }
 
-  // function improves estimate for expected size of message
-  void hint_size() {
-    if (m_exp_len != NULL) {
-      size_t size = m_data.size();
-      if (size > __atomic_load_n(m_exp_len, __ATOMIC_RELAXED)) {
-        //log larger then expected, just expand
-        __atomic_store_n(m_exp_len, size + 10, __ATOMIC_RELAXED);
-      }
-      else {
-        //asymptotically adapt expected size to message size
-        __atomic_store_n(m_exp_len, (size + 10 + m_buf_len*31) / 32, __ATOMIC_RELAXED);
-      }
-    }
+  std::string_view strv() const override {
+    return cos.get_stream().strv();
   }
-
-  void set_str(const std::string &s) {
-    get_ostream() << s;
+  std::size_t size() const override {
+    return cos.get_stream().strv().size();
   }
 
-  std::string get_str() const {
-    return m_data.get_str();
-  }
+private:
+  CachedStackStringStream cos;
+};
 
-  // returns current size of content
-  size_t size() const {
-    return m_data.size();
-  }
+class ConcreteEntry : public Entry {
+public:
+  static constexpr std::size_t buffer_size = 4096;
 
-  // extracts up to avail chars of content
-  int snprintf(char* dst, size_t avail) const {
-    return m_data.snprintf(dst, avail);
+  ConcreteEntry() = delete;
+  ConcreteEntry(const Entry& e) : Entry(e) {
+    auto strv = e.strv();
+    str.reserve(strv.size());
+    str.insert(str.end(), strv.begin(), strv.end());
   }
-
-  void finish() {
-    m_streambuf->finish();
+  ConcreteEntry& operator=(const Entry& e) {
+    Entry::operator=(e);
+    auto strv = e.strv();
+    str.clear();
+    str.reserve(strv.size());
+    str.insert(str.end(), strv.begin(), strv.end());
```

Comment:
can use `str.assign(strv.begin(), strv.end());` instead of `clear()` + `insert()`

Reply:
thanks